### ISSUE TITLE
Simple Object Creation with Template String - incorrect (and missing) compiler errors

### ISSUE THREAD

**Issue by viyaha** [microsoft/TypeScript#55634]
### 🔎 Search Terms

template string object key 
template string

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.2.2#code/JYOwLgpgTgZghgYwgAgEoQI4FdhQgEwAUoB7AB2jAE9kBvAWAChkXlIBnMARgC5kQsAWwBG0ANxMAvkyahIsRCgDyZMMBIg4AG2LlKNBs1YduAfj4CR4qTMYwsIBGo3JQ+CAA8AEnHYA5EjAAWTgwBAALUABzABUqCgAKN08LIVEoAEo6JlZkAHo85ABJRxIoPCdkAFFysoBaIIh2djgolDqAPmRPCido5BbBFGhSKGRBdijkX2QIuBAo-rBwlDJSCihqZBd57o9gTn6SYQArHNYEDU5kPAwlU750bFwCXQ2tgF5so1yWAG0AAYmAAktGSHkkAIAunwAEwAGnOLGkjCR+Tyl3KEEqIzKaNu9xOgJBYJA7gh0OQX1hYmQtlyBQG4RIWC0+GQommIGqtSgaMuIGu5DAhL4Kmcmh0630VO+v1YxKaYFB4MhMOQsLRKNyaMZmIqYG6vLRwsJis4KrJnjVsppNlRdgcfRc4J87BCYUiCziiXBfC4WUMDMK7GZrPZAHcygBrflXQ0Eh5oTA4PBEaWbGhfIPy5Dm5Wk8lquGIn7IFG6wr67FgfGYM1ApWWouU6kSB25AVC1Si5Di9SSt4y7MVst6soGk090755vW1sa9ujoA

### 💻 Code

```ts
interface RequiredProperty {
    test1: number;
}

interface OptionalProperty {
    test1?: number;
}

function indexHasNotMatchingType(index: number) {
    // Incorrect Error-Message -> expecting same error msg as changing the property on an existing obj
    const reqObj: RequiredProperty = {
        [`test${index}`]: 2,
    }

    //correct error
    reqObj[`test${index}`] = 2; 

    // should be an Error
    const optObj: OptionalProperty = {
        [`test${index}`]: 2
    }
    
    // correct error
    optObj[`test${index}`] = 2;
}

function indexHasMatchingType(index: 1) {
    // should work
    const reqObj: RequiredProperty = {
        [`test${index}`]: 2,
    }

    // correct
    reqObj[`test${index}`] = 2;

    const optObj: OptionalProperty = {}

    // correct
    optObj[`test${index}`] = 2;
}


```

### 🙁 Actual behavior

3 Issues here:

1) (minor issue, can be ignored most likely) Trying to create an object with a not matching template string key with an required property should fail with the same err-msg as trying to change the same key afterwards

2)  Object with optional property can be created with template string key that does not match the index signatur of the interface. (Property changes afterwards fails correctly)

3) Object with required property cannot be created with correct existing Temaplte string key. (Property of existing Obj can be changed)

### 🙂 Expected behavior

1) Same error for creationg and updating an obj with an template string key.
2) User should not be able to create an object with an incorrect template string key if the properties are optional
3) User should be able to create an object with an template string key and not only change it after creation.

**Comment by Andarist** [comment]
1. The context at the assignment site and the context at indexed access are different. The same error message wouldn't fit both sites. The right side of the assignment isn't "yet" of the declared type - it can have a different type and thus the error doesn't refer to that declared type in a way the other site can refer to it.
2. You can always assign objects with *more* properties to an object with *less* properties (as long as the overlapping properties of the target and the source have matching types). What you likely expected here is the "excess property check". It's more like a lint rule than a type error really and computed properties are exempted from this check. This assignment is valid so the lack of error there is expected.
3. This is how computed properties work - based on heuristics the checker doesn't assume that you have wanted a concrete property there if you have used a template expression. It gets widened. To prevent that you can `as const` your expression
